From ff1257383c23f30df8e7ed41da38e08e130956d1 Mon Sep 17 00:00:00 2001
From: John Greene <jogreene@redhat.com>
Date: Wed, 10 Sep 2014 18:28:54 -0400
Subject: [ethernet] ixgbe: fix ixgbe_stop_mac_link_on_d3_82599 to check mng correctly

Message-id: <1410373750-11341-35-git-send-email-jogreene@redhat.com>
Patchwork-id: 93557
O-Subject: [RHEL7.1 PATCH 34/50] ixgbe: fix ixgbe_stop_mac_link_on_d3_82599 to check mng correctly
Bugzilla: 1091123
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Neil Horman <nhorman@redhat.com>

BZ: 1091123
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7919168
Tested: In beaker. See BZ for details

Previously, we did a full check to see if MNG FW was running. Instead,
we should only check to see whether it could be enabled. Since it may
become active while down, we don't want to bring the link down.

CC: Arun Sharma <asharma@fb.com>
Signed-off-by: Don Skidmore <donald.c.skidmore@intel.com>
Signed-off-by: Jacob Keller <jacob.e.keller@intel.com>
Tested-by: Phil Schmitt <phillip.j.schmitt@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
(cherry picked from commit 8036d29f409f6a1d54ac409ede273197c4914e3f)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c b/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
index 2f6c26f..de94b5e 100644
--- a/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
+++ b/drivers/net/ethernet/intel/ixgbe/ixgbe_82599.c
@@ -517,12 +517,16 @@ out:
  **/
 static void ixgbe_stop_mac_link_on_d3_82599(struct ixgbe_hw *hw)
 {
- u32 autoc2_reg;
+ u32 autoc2_reg, fwsm;
  u16 ee_ctrl_2 = 0;
 
  hw->eeprom.ops.read(hw, IXGBE_EEPROM_CTRL_2, &ee_ctrl_2);
 
- if (!hw->mng_fw_enabled && !hw->wol_enabled &&
+ /* Check to see if MNG FW could be enabled */
+ fwsm = IXGBE_READ_REG(hw, IXGBE_FWSM);
+
+ if (((fwsm & IXGBE_FWSM_MODE_MASK) != IXGBE_FWSM_FW_MODE_PT) &&
+     !hw->wol_enabled &&
      ee_ctrl_2 & IXGBE_EEPROM_CCD_BIT) {
   autoc2_reg = IXGBE_READ_REG(hw, IXGBE_AUTOC2);
   autoc2_reg |= IXGBE_AUTOC2_LINK_DISABLE_ON_D3_MASK;
-- 
1.7.1